datafusion/optimizer/src/decorrelate_where_exists.rs

```diff
@@ -0,0 +1,558 @@
+    #[test]
+    fn exists_subquery_where_not_eq() -> Result<()> {
+        let sq = Arc::new(
+            LogicalPlanBuilder::from(scan_tpch_table("orders"))
+                .filter(col("customer.c_custkey").not_eq(col("orders.o_custkey")))?
+                .project(vec![col("orders.o_custkey")])?
+                .build()?,
+        );
+
+        let plan = LogicalPlanBuilder::from(scan_tpch_table("customer"))
+            .filter(exists(sq))?
+            .project(vec![col("customer.c_custkey")])?
+            .build()?;
+
+        let expected = r#"cannot optimize non-correlated subquery"#;
+
+        assert_optimizer_err(&DecorrelateWhereExists::new(), &plan, expected);
+        Ok(())
+    }
+
+    /// Test for correlated exists subquery less than
+    #[test]
+    fn exists_subquery_where_less_than() -> Result<()> {
+        let sq = Arc::new(
+            LogicalPlanBuilder::from(scan_tpch_table("orders"))
+                .filter(col("customer.c_custkey").lt(col("orders.o_custkey")))?
+                .project(vec![col("orders.o_custkey")])?
+                .build()?,
+        );
+
+        let plan = LogicalPlanBuilder::from(scan_tpch_table("customer"))
+            .filter(exists(sq))?
+            .project(vec![col("customer.c_custkey")])?
+            .build()?;
+
+        let expected = r#"can't optimize < column comparison"#;
+
+        assert_optimizer_err(&DecorrelateWhereExists::new(), &plan, expected);
+        Ok(())
+    }
+
+    /// Test for correlated exists subquery filter with subquery disjunction
+    #[test]
+    fn exists_subquery_with_subquery_disjunction() -> Result<()> {
+        let sq = Arc::new(
+            LogicalPlanBuilder::from(scan_tpch_table("orders"))
+                .filter(
+                    col("customer.c_custkey")
+                        .eq(col("orders.o_custkey"))
+                        .or(col("o_orderkey").eq(lit(1))),
+                )?
+                .project(vec![col("orders.o_custkey")])?
+                .build()?,
+        );
+
+        let plan = LogicalPlanBuilder::from(scan_tpch_table("customer"))
+            .filter(exists(sq))?
+            .project(vec![col("customer.c_custkey")])?
+            .build()?;
+
+        let expected = r#"cannot optimize correlated disjunctions"#;
+
+        assert_optimizer_err(&DecorrelateWhereExists::new(), &plan, expected);
+        Ok(())
+    }
+
+    /// Test for correlated exists without projection
+    #[test]
+    fn exists_subquery_no_projection() -> Result<()> {
+        let sq = Arc::new(
+            LogicalPlanBuilder::from(scan_tpch_table("orders"))
+                .filter(col("customer.c_custkey").eq(col("orders.o_custkey")))?
+                .build()?,
+        );
+
+        let plan = LogicalPlanBuilder::from(scan_tpch_table("customer"))
+            .filter(exists(sq))?
+            .project(vec![col("customer.c_custkey")])?
+            .build()?;
+
+        let expected = r#"Could not coerce into Filter!"#;
```

Comment:
I think this error would be hard to understand as a user, for what it is worth

Reply:
`DataFusionError::Context` told the user that it was in `decorrelate_where_exists.rs` and some other helpful info, but I'm happy to remove it and update the expectations. I'll go ahead and do that.

Reply:
This way is fine too

Reply:
The assertion matches any subset of the error message. I updated it to `cannot optimize non-correlated subquery`, the full message is:

```
cannot optimize non-correlated subquery at datafusion/optimizer/src/decorrelate_where_exists.rs:141
caused by
Error during planning: Could not coerce into Filter! at datafusion/expr/src/logical_plan/plan.rs:1117
```